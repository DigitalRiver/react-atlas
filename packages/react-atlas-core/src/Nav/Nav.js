import React, { cloneElement } from "react";
import PropTypes from "prop-types";

class Nav extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      "activeIndex": props.activeIndex,
      "collapsed": props.collapsed
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.activeIndex !== this.props.activeIndex) {
      this.setState({ "activeIndex": nextProps.activeIndex });
    }
  }

  _handleClick = (index, event) => {
    this.setState({ "activeIndex": index });
    if (this.props.onClick) {
      this.props.onClick(index, event);
    }
  };

  _handleCollapse = (index, event) => {
    this.setState({ "collapsed": !this.state.collapsed });
    if (this.props.onClick) {
      this.props.onClick(index, event);
    }
  };

  render() {
    const { className, style, children, subNav } = this.props;
    return (
      <ul styleName="nav" className={className} style={style}>
        {React.Children.map(children, (child, index) => {
          let active = false;
          if (
            child.props.active ||
            child.props.navKey === this.state.activeIndex
          ) {
            active = true;
          }
          if (child.type.name === "Nav") {
            return cloneElement(child, {
              "activeIndex": this.state.activeIndex,
              "subNav": true,
              "onClick": this._handleClick
            });
          }
          let isParent = subNav && index === 0;
          return cloneElement(child, {
            "subNav": subNav && !isParent,
            "parent": isParent,
            "collapsed": this.state.collapsed,
            active,
            "onClick": isParent ? this._handleCollapse : this._handleClick
          });
        })}
      </ul>
    );
  }
}

Nav.propTypes = {
  /**
   * Define the selected NavItem component in the Nav.
   */
  "activeIndex": PropTypes.number,
  /**
   * Child elements, typically NavItem component or Nav for grouping sub-NavItems.
   */
  "children": PropTypes.node.isRequired,

  /** An object, array, or string of CSS classes to apply to Nav.*/
  "className": PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),
  /**
   * Define Nav container is collapsed or not when initialization, normally used in sub Nav.
   */
  "collapsed": PropTypes.bool,

  /**
   * Function that will be executed when onClick event occurs.
   * @examples '<Nav onClick={onClickHandler}/>'
   */
  "onClick": PropTypes.func,

  /**
   * Pass inline styling here.
   */
  "style": PropTypes.object,

  /**
   * Set true If is a sub-Nav container.
   *
   * @ignore
   */
  "subNav": PropTypes.bool
};

export default Nav;
