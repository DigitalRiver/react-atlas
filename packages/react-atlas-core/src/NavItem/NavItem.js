import React from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import { ButtonCore } from "./../Button";

class NavItem extends React.PureComponent {
  _handleClick = event => {
    if (this.props.onClick) {
      this.props.onClick(this.props.navKey, event);
    }
  };

  render() {
    const {
      className,
      style,
      active,
      disabled,
      href,
      parent,
      subNav,
      collapsed,
      children
    } = this.props;
    if (subNav && collapsed) {
      return "";
    }

    return (
      <li
        role="presentation"
        styleName={cx("navItem", { active, disabled, collapsed })}
        style={style}
        className={className}
      >
        <ButtonCore
          disabled={disabled}
          styleName={cx("link", { disabled, subNav })}
          href={href}
          onClick={this._handleClick}
        >
          {children}
          <i styleName={cx({ "caret": parent })} />
        </ButtonCore>
      </li>
    );
  }
}

NavItem.propTypes = {
  /**
   * Child elements, plain text or HTML/React element.
   */
  "children": PropTypes.node.isRequired,

  /** An object, array, or string of CSS classes to apply to NavItem.*/
  "className": PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),

  /**
   * When true, NavItem will be disabled.
   */
  "disabled": PropTypes.bool,

  /**
   * The URL that the NavItem will link to.
   */
  "href": PropTypes.string,

  /**
   * Specify NavItem unique key
   */
  "navKey": PropTypes.number.isRequired,
  /**
   * Pass inline styling here.
   */
  "style": PropTypes.object,
  /**
   * When true, NavItem will be set as active.
   *
   * @ignore
   */
  "active": PropTypes.bool,
  /**
   * Only first NavItem in sub-Nav container will be true.
   *
   * @ignore
   */
  "parent": PropTypes.bool,
  /**
   * Each NavItem in sub-Nav container will be true except first one.
   *
   * @ignore
   */
  "subNav": PropTypes.bool,
  /**
   * Only use in sub Nav container, defines parent NavItem caret and subNav show or not.
   *
   * @ignore
   */
  "collapsed": PropTypes.bool,
  /**
   * callback for Nav Component
   *
   * @ignore
   */
  "onClick": PropTypes.func
};

NavItem.defaultProps = {
  "className": ""
};

export default NavItem;
