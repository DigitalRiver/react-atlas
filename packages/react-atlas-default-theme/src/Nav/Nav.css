.nav {
    padding-left: 0;
    margin-bottom: 0;
    list-style: none;
}
