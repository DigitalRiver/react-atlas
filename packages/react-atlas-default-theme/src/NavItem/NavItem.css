@import '../Button/Button.css';

.navItem button {
    width: 100%;
    border-color: #eee;
    border-left: solid 5px transparent;
    color: #555;
    position: relative;
    display: block;
    padding: 10px 15px;
    background-color: #fff;
    text-align: left;
}

.navItem button:hover {
    background-color: #eee;
}

.active > button {
    border-left-color: #003058;
    background-color: #eee;
    color: #003058;
}

.base {
    border-style: none;
    border-top: 1px solid #ccc;
}

.navItem .caret {
    position: relative;
    display: inline-block;
    width: 18px;
    height: 10px;
    vertical-align: middle;
    float: right;
    clear: right;
    border-color: #555;
    transform: none;
    margin: 5px 0 0;
    transition: transform 300ms;
}

.collapsed .caret {
    transform: rotate(90deg);
    transition: transform 300ms;
}

.navItem .caret::before {
    content: "";
    border: solid #003058;
    border-width: 0 2px 2px 0;
    display: inline-block;
    padding: 5px;
    transform: rotate(45deg);
    position: absolute;
    top: -5px;
    left: 5px;
}

.navItem .subNav {
    padding-left: 30px;
}