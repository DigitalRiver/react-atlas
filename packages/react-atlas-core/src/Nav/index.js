import { default as NavCore } from "./Nav";
export { NavCore };
