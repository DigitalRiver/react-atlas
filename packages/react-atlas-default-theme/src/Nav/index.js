import * as NavStyle from "./Nav.css";
export { NavStyle };
