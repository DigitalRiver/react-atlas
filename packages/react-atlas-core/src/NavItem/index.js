import { default as NavItemCore } from "./NavItem";
export { NavItemCore };
