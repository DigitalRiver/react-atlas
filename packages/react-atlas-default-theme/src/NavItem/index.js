import * as NavItemStyle from "./NavItem.css";
export { NavItemStyle };
