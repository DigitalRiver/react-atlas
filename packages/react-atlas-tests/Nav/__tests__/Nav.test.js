import React from "react";
import { mount } from "enzyme";
import { NavCore } from "../../../react-atlas-core/src/Nav/index";
import { NavItemCore } from "../../../react-atlas-core/src/NavItem/index";
import renderer from "react-test-renderer";

describe("Test Nav component render", () => {
  it("Render correctly", () => {
    const tree = renderer
      .create(
        <NavCore>
          <NavItemCore navKey={0}>HOME</NavItemCore>
          <NavItemCore navKey={1}>Catalog</NavItemCore>
          <NavItemCore navKey={2}>Customer Service</NavItemCore>
          <NavItemCore navKey={3}>Administration</NavItemCore>
          <NavItemCore navKey={4}>Support</NavItemCore>
        </NavCore>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const _findItem = (node, navKey) =>
  node.name() === "NavItem" && node.props().navKey === navKey;

const _findActiveItem = node =>
  node.name() === "NavItem" && node.props().active;

describe("Testing Nav component", () => {
  it("Nav - Simple default select NavItem", function() {
    const result = mount(
      <NavCore activeIndex={1}>
        <NavItemCore navKey={0}>HOME</NavItemCore>
        <NavItemCore navKey={1}>Catalog</NavItemCore>
        <NavItemCore navKey={2}>Customer Service</NavItemCore>
        <NavItemCore navKey={3}>Administration</NavItemCore>
        <NavItemCore navKey={4}>Support</NavItemCore>
      </NavCore>
    );
    expect(result.findWhere(n => _findActiveItem(n)).props().navKey).toEqual(1);
  });
  it("Nav - Default select inner NavItem", function() {
    const result = mount(
      <NavCore activeIndex={3.1}>
        <NavItemCore navKey={0}>HOME</NavItemCore>
        <NavItemCore navKey={1}>Catalog</NavItemCore>
        <NavCore>
          <NavItemCore navKey={3}>Reports</NavItemCore>
          <NavItemCore navKey={3.1}>Create a Report</NavItemCore>
          <NavItemCore navKey={3.2}>Edit a Report</NavItemCore>
        </NavCore>
        <NavItemCore navKey={4}>Administration</NavItemCore>
        <NavItemCore navKey={5}>Support</NavItemCore>
      </NavCore>
    );
    expect(result.findWhere(n => _findActiveItem(n)).props().navKey).toEqual(
      3.1
    );
  });
  it("Nav - Click to changing active NavItem", function() {
    const result = mount(
      <NavCore activeIndex={1}>
        <NavItemCore navKey={0}>HOME</NavItemCore>
        <NavItemCore navKey={1}>Catalog</NavItemCore>
        <NavItemCore navKey={2}>Customer Service</NavItemCore>
        <NavCore>
          <NavItemCore navKey={3}>Reports</NavItemCore>
          <NavItemCore navKey={3.1}>Create a Report</NavItemCore>
          <NavItemCore navKey={3.2}>Edit a Report</NavItemCore>
        </NavCore>
        <NavItemCore navKey={4}>Support</NavItemCore>
      </NavCore>
    );
    result
      .findWhere(n => _findItem(n, 2))
      .find("button")
      .simulate("click");
    expect(result.findWhere(n => _findActiveItem(n)).props().navKey).toEqual(2);

    result
      .findWhere(n => _findItem(n, 3.1))
      .find("button")
      .simulate("click");
    expect(result.findWhere(n => _findActiveItem(n)).props().navKey).toEqual(
      3.1
    );
  });
});
