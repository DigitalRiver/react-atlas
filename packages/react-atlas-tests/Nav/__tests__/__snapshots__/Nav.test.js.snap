// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Nav component render Render correctly 1`] = `
<ul
  className={undefined}
  style={undefined}
  styleName="nav"
>
  <li
    className=""
    role="presentation"
    style={undefined}
    styleName="navItem"
  >
    <button
      className=""
      href={undefined}
      id={undefined}
      name={undefined}
      onClick={[Function]}
      style={undefined}
      styleName="button uppercase default_btn default_btn_hover"
      tabIndex={0}
      type="button"
    >
      HOME
      <i
        styleName=""
      />
    </button>
  </li>
  <li
    className=""
    role="presentation"
    style={undefined}
    styleName="navItem"
  >
    <button
      className=""
      href={undefined}
      id={undefined}
      name={undefined}
      onClick={[Function]}
      style={undefined}
      styleName="button uppercase default_btn default_btn_hover"
      tabIndex={0}
      type="button"
    >
      Catalog
      <i
        styleName=""
      />
    </button>
  </li>
  <li
    className=""
    role="presentation"
    style={undefined}
    styleName="navItem"
  >
    <button
      className=""
      href={undefined}
      id={undefined}
      name={undefined}
      onClick={[Function]}
      style={undefined}
      styleName="button uppercase default_btn default_btn_hover"
      tabIndex={0}
      type="button"
    >
      Customer Service
      <i
        styleName=""
      />
    </button>
  </li>
  <li
    className=""
    role="presentation"
    style={undefined}
    styleName="navItem"
  >
    <button
      className=""
      href={undefined}
      id={undefined}
      name={undefined}
      onClick={[Function]}
      style={undefined}
      styleName="button uppercase default_btn default_btn_hover"
      tabIndex={0}
      type="button"
    >
      Administration
      <i
        styleName=""
      />
    </button>
  </li>
  <li
    className=""
    role="presentation"
    style={undefined}
    styleName="navItem"
  >
    <button
      className=""
      href={undefined}
      id={undefined}
      name={undefined}
      onClick={[Function]}
      style={undefined}
      styleName="button uppercase default_btn default_btn_hover"
      tabIndex={0}
      type="button"
    >
      Support
      <i
        styleName=""
      />
    </button>
  </li>
</ul>
`;
